File: backend/src/main/java/io/github/xpakx/ladder/user/dto/UserWithData.java
package io.github.xpakx.ladder.user.dto;

import io.github.xpakx.ladder.collaboration.dto.CollabTaskDetails;
import io.github.xpakx.ladder.collaboration.dto.CollaborationWithProject;
import io.github.xpakx.ladder.filter.dto.FilterDetails;
import io.github.xpakx.ladder.habit.dto.HabitCompletionDetails;
import io.github.xpakx.ladder.habit.dto.HabitDetails;
import io.github.xpakx.ladder.label.dto.LabelDetails;
import io.github.xpakx.ladder.project.dto.ProjectDetails;
import io.github.xpakx.ladder.task.dto.TaskDetails;
import lombok.Getter;
import lombok.Setter;

import java.util.List;

@Setter
@Getter
public class UserWithData {
    private Integer id;
    private String username;
    private boolean projectCollapsed;
    private List<ProjectDetails> projects;
    private List<TaskDetails> tasks;
    private List<LabelDetails> labels;
    private List<HabitDetails> habits;
    private List<HabitCompletionDetails> todayHabitCompletions;
    private List<FilterDetails> filters;
    private List<CollaborationWithProject> collabs;
    private List<CollabTaskDetails> collabTasks;
}

File: frontend/src/app/sync/dto/user-with-data.ts
import { CollabProjectData } from "src/app/project/dto/collab-project-data";
import { CollabTaskDetails } from "src/app/task/dto/collab-task-details";
import { FilterDetails } from "src/app/filter/dto/filter-details";
import { HabitCompletionDetails } from "src/app/habit/dto/habit-completion-details";
import { HabitDetails } from "src/app/habit/dto/habit-details";
import { LabelDetails } from "src/app/label/dto/label-details";
import { ProjectDetails } from "src/app/project/dto/project-details";
import { TaskDetails } from "src/app/task/dto/task-details";

export interface UserWithData {
    username: string;
    id: number;
    projectCollapsed: boolean;
    projects: ProjectDetails[];
    tasks: TaskDetails[];
    labels: LabelDetails[];
    habits: HabitDetails[];
    todayHabitCompletions: HabitCompletionDetails[];
    filters: FilterDetails[];
    collabs: CollabProjectData[];
    collabTasks: CollabTaskDetails[];
}